### PR TITLE
New: Adds SendWP integration to email settings admin page.

### LINKED ISSUE #6229: [Existing functionality] should integrate with [some application]
## User Story

As a SendWP user, I want to see an integration with SendWP so that SendWP and Give customers can automatically sign up if needed.

## Details

Other plugins in the space allow users to sign up for SendWP within their app when they have trouble with email deliverability. I believe on the email settings page would be a good place to integrate.

## Additional Context

To be clear, I own SendWP. 

## Acceptance Criteria

-   [ ] Can automatically download and connect SendWP by clicking button.
-   [ ] Can disconnect if needed.

### PR DESCRIPTION
**Pull request by mrpritchett**
Resolves #6229 

## Description

This PR adds the ability to automatically download the SendWP plugin, register and attempt to activate the plugin. It includes a description of what SendWP is, when Give recommends it, and why you should use it.

## Affects

GiveWP Email Settings page

## Visuals

Screen Shot 2022-02-12 at 1 25 42 PM
Screen Shot 2022-02-12 at 1 28 51 PM

## Testing Instructions

- This can only be tested from a publicly available URL (not local) as the process of connecting & disconnecting requires redirects back to the site.

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit and/or end-to-end tests
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
Hi @mrpritchett!

Thank you for taking the time to make an Issue and PR for integrating with SendWP. That's awesome and should exist.

I don't believe the best place to do this, however, is directly within GiveWP core. This would make a better add-on plugin that users can install alongside GiveWP and SendWP. Either that or you can choose to put the integration right into SendWP. Having an additional add-on, however, gives you a marketing opportunity and scopes support for the plugin to just the integration.

As such, I'm going to close this and the Issue, but encourage you to go ahead with one of the two options. If you need some sort of additional hook within GiveWP in order for the integration to work, feel free to open up an Issue/PR for that hook. If you have any questions on the integration, go ahead and [start a discussion](https://github.com/impress-org/givewp/discussions).

Cheers!

**Comment by kjohnson** [comment]
@JasonTheAdams this may be a strategic partnership opportunity for Give. I think Devin and Matt are already considering this. The nature of SendWP lends itself to having an upsell in-plugin with a hard-coded affiliate code,

**Comment by mrpritchett** [comment]
@kbjohnson90 - Thanks for commenting. I was just getting ready to say the same thing! :) 

We're happy to go with whatever is best for GiveWP and your customers, but I had spoken with Devin about the integration a few months ago. We typically use this integration with our partners, but are happy to go the add-on route if required. As Kyle mentioned, we typically integrate directly in the plugin as this helps ensure the benefit for GiveWP and SendWP's customers. Let me know what you and Devin would like us to do and I'm happy to oblige. 

Thanks again for the awesome work!

**Comment by JasonTheAdams** [comment]
Apparently there was some conversations I wasn't aware about. Hahah! I'm re-opening this as we talk more about this internally and such.

File: includes/admin/misc-functions.php [hunk 1]
```diff
@@ -249,3 +249,99 @@ function give_add_ons_feed( $feed_type = '', $echo = true ) {
...
+add_action( 'wp_ajax_give_sendwp_remote_install', 'give_sendwp_remote_install_handler' );
+
+/**
+ * Handle deactivation of SendWP via ajax
+ *
+ * @since 2.9.15
+ */
+function give_sendwp_disconnect () {
+
+	if ( ! current_user_can( 'manage_shop_settings' ) ) {
```
Comment (Benunc) [inline includes/admin/misc-functions.php:335]: Is that a GiveWP function? seems like EDD or something.

File: includes/admin/settings/class-settings-email.php [hunk 2]
```diff
@@ -36,6 +36,148 @@ public function __construct() {
...
+                <?php
+            else :
+                ?>
+                <tr valign="top" <?php echo ! empty( $field['wrapper_class'] ) ? 'class="' . $field['wrapper_class'] . '"' : ''; ?>>
+                    <th scope="row" class="titledesc">
+                        <label for="<?php echo esc_attr( $field['id'] ); ?>"><?php echo esc_html( $field['name'] ); ?></label>
+                    </th>
+                    <td class="give-forminp">
+                        <div class="give-field-description">
+                            <?php _e( 'We recommend SendWP to ensure quick and reliable delivery of all emails sent from your site, such as donation receipts, recurring donation renewal reminders, password resets, and more.', 'give' ); ?> <?php printf( __( '%sLearn more%s', 'give' ), '<a href="https://sendwp.com/" target="_blank" rel="noopener noreferrer">', '</a>' ); ?>
```
Comment (Benunc) [inline includes/admin/settings/class-settings-email.php:92]:
```suggestion
                            <?php _e( 'GiveWP recommends SendWP to ensure quick and reliable delivery of all emails sent from your site, such as donation receipts, recurring donation renewal reminders, password resets, and more.', 'give' ); ?> <?php printf( __( '%sLearn more%s', 'give' ), '<a href="https://go.givewp.com/sendwpinternal" target="_blank" rel="noopener noreferrer">', '</a>' ); ?>
```

File: includes/admin/settings/class-settings-email.php [hunk 3]
```diff
@@ -84,6 +226,12 @@ public function get_settings() {
 							'default' => get_bloginfo( 'admin_email' ),
 							'type'    => 'text',
 						),
+                        array(
+							'id'      => 'sendwp',
+							'name'    => esc_html__( 'SendWP', 'give' ),
+							'desc'    => esc_html__( 'We recommend SendWP to ensure quick and reliable delivery of all emails sent from your store, such as donation receipts, recurring donation renewal reminders, password resets, and more.', 'give' ),
```
Comment (Benunc) [inline includes/admin/settings/class-settings-email.php:232]:
```suggestion
							'desc'    => esc_html__( 'GiveWP recommends SendWP to ensure quick and reliable delivery of all emails sent from your site, such as donation receipts, recurring donation renewal reminders, password resets, and more.', 'give' ),
```

**Review by Benunc — COMMENTED** [review]
I am still testing functionality of the PR, these are changes that need to be made to contextual docs in the meantime.

**Comment by Benunc** [comment]
Once I'm logged into my account, I get a notice about SendWP needing to be activated. When I click the link, I'm getting a server error: 

CleanShot 2022-04-25 at 16 44 07

**Comment by mrpritchett** [comment]
Hey @Benunc - Matt Pritchett here, if you don't mind letting me know when your initial review is complete, I'll jump into resolving these issues. But I didn't want to step on any toes in the meantime. Feel free to let me know if another workflow or process works for you. I'm happy to adjust. 😃

File: includes/admin/misc-functions.php (hunk 1 above)
Comment (mrpritchett) [reply includes/admin/misc-functions.php]: I've updated this to be more in line with GiveWP capabilities. Good catch!

File: includes/admin/settings/class-settings-email.php (hunk 3 above)
Comment (mrpritchett) [reply includes/admin/settings/class-settings-email.php:232]: This has been updated.

**Comment by mrpritchett** [comment]
> When clicking to connect, it briefly shows "undefined" on the button before the redirection to sendwp CleanShot 2022-04-25 at 16 18 42

I've updated the loading state for both connect and disconnect states to not conflict as it was before.

**Comment by mrpritchett** [comment]
Hey Ben, can you provide me with any environment details for your test environment (PHP Version, plugins active, etc)? I can see the 500 on our end in our logs, but the error messaging is basically nonexistent. I'm going to continue tracing it in the meantime, but those details would be helpful. Thanks!

**Comment by mrpritchett** [comment]
@Benunc - I realized I didn't tag you (not 100% sure what that procedure is for alerting contributors), but please see my above comment when you have a few minutes. I mainly looking for what the error was on your staging server, as that is where the error occurred (http post request from our server to your site resulted in a destination server error, causing an uncaught error on our side, which we're going to fix). 

Just as a heads up, I'm on a team retreat this week, so the work will be a little slower on our end than normal. 

Thanks again for reviewing this and let me know if I need to change anything about my responses or workflow.

**Comment by mrpritchett** [comment]
Hey @Benunc - Kyle was able to get me a similar sandbox to the one you used and I was able to connect through both an existing login and a new registration without error. Any chance you could provide more context for the error you experienced?

**Comment by Benunc** [comment]
Hey @mrpritchett sorry for ghosting you here. I have all my GitHub notifications turned off because I am in here so infrequently. 

I just spun up a new sandbox and tried connecting again, and this time it went through with no issue.

I think this is ready for primetime.

**Comment by mrpritchett** [comment]
No worries @Benunc. Let me know what you and the team need from me/us to make this happen.

**Review by Benunc — APPROVED** [review]
This has been tested and the functionality works to connect to SendWP, and all automated QA tests pass.

**Comment by ravinderk** [comment]
The team decided to hold on merging this until our next minor release.

**Pull request merged**